forge/src/main/java/me/lucko/luckperms/forge/ForgeClassPathAppender.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge;
+
+import cpw.mods.cl.ModuleClassLoader;
+import cpw.mods.modlauncher.Launcher;
+import cpw.mods.modlauncher.TransformingClassLoader;
+import me.lucko.luckperms.common.plugin.classpath.ClassPathAppender;
+import me.lucko.luckperms.common.plugin.logging.PluginLogger;
+import sun.misc.Unsafe;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+public class ForgeClassPathAppender extends URLClassLoader implements ClassPathAppender {
+    private final PluginLogger logger;
+    private final Set<String> invalidClasses;
+    private final Map<String, ClassLoader> parentLoaders;
+
+    public ForgeClassPathAppender(PluginLogger logger) {
+        super(new URL[0], (ModuleClassLoader) Launcher.class.getClassLoader());
+        this.logger = logger;
+        this.invalidClasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        this.parentLoaders = getParentLoaders((TransformingClassLoader) Thread.currentThread().getContextClassLoader());
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        if (invalidClasses.contains(name)) {
+            throw new ClassNotFoundException(name);
+        }
+
+        invalidClasses.add(name);
+        Class<?> validClass = super.findClass(name);
+        invalidClasses.remove(name);
+        return validClass;
+    }
+
+    @Override
+    public void addJarToClasspath(Path file) {
+        // For debugging purposes
+        // this.logger.info("Adding " + file + " to the classpath");
+
+        try {
+            addURL(file.toUri().toURL());
+        } catch (MalformedURLException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        if (this.parentLoaders == null) {
+            return;
+        }
+
+        try (JarFile jarFile = new JarFile(file.toFile(), false)) {
+            for (Enumeration<JarEntry> enumeration = jarFile.entries(); enumeration.hasMoreElements(); ) {
+                JarEntry jarEntry = enumeration.nextElement();
+                if (jarEntry.isDirectory() || !jarEntry.getName().endsWith(".class")) {
```

Comment:
I don't think this approach will work with native libraries or other required resources?

I'm pretty sure some of the JDBC drivers need that

Reply:
The parentLoaders is used by modlauncher for loading classes not resources.

Resources including JDBC drivers are loaded fine (Tested H2 & MySQL) as they call `getResourceAsStream` which uses the ClassLoader which loaded the class, in the case of H2 it's the `IsolatedClassLoader` and for MySQL it's the `ForgeClassPathAppender` as these ClassLoaders extend URLClassLoader they themselves handling pulling the resources from the jars.